src/mscorlib/src/System/Globalization/EncodingTable.Unix.cs

```diff
@@ -1,101 +1,187 @@
+                }
+                else
+                {
+                    //The name that we're looking for is greater than our current index
+                    left = index;
+                }
+            }
 
-                default:
-                    return CodePageUtf8;
+            //Walk the remaining elements (it'll be 3 or fewer).
+            for (; left <= right; left++)
+            {
+                InternalEncodingDataItem dataItem = encodingDataTable[left];
+                if (string.Compare(name, dataItem.WebName, StringComparison.OrdinalIgnoreCase) == 0)
+                {
+                    return dataItem.CodePage;
+                }
             }
+            // The encoding name is not valid.
+            throw new ArgumentException(
+                String.Format(
+                    CultureInfo.CurrentCulture,
+                    Environment.GetResourceString("Argument_EncodingNotSupported"), name), "name");
         }
-    
+
         [System.Security.SecuritySafeCritical]  // auto-generated
-        unsafe internal static CodePageDataItem GetCodePageDataItem(int codepage) {
-            // TODO: Implement this fully.
-            switch (codepage)
+        private static CodePageDataItem InternalGetCodePageDataItem(int codepage)
+        {
+            for (int i = 0; i < encodingDataTableItems.Length; i++)
             {
-                case CodePageUtf7:
-                    return new CodePageDataItem("utf-7", CodePageUtf7, "utf-7", "utf-7", 771);
+                InternalCodePageDataItem dataItem = encodingDataTableItems[i];
+                if (dataItem.CodePage == codepage)
+                {
+                    return new CodePageDataItem(dataItem.Name, dataItem.UiFamilyCodePage, dataItem.Flags);
+                }
+            }
+
+            return null;
+        }
 
-                case CodePageUtf8:
-                    return new CodePageDataItem("utf-8", CodePageUtf8, "utf-8", "utf-8", 771);
+        // PAL ends here.
 
-                case CodePageUtf16:
-                    return new CodePageDataItem("utf-16", CodePageUtf16, "utf-16", "utf-16", 771);
+        private struct InternalEncodingDataItem
+        {
+            public InternalEncodingDataItem(string webName, UInt16 codePage)
+            {
+                WebName = webName;
+                CodePage = codePage;
+            }
 
-                case CodePageUtf16BE:
-                    return new CodePageDataItem("utf-16BE", CodePageUtf16BE, "utf-16BE", "utf-16BE", 771);
+            public string WebName { get; private set; }
+            public UInt16 CodePage { get; private set; }
+        }
 
-                case CodePageUtf32:
-                    return new CodePageDataItem("utf-32", CodePageUtf32, "utf-32", "utf-32", 771);
+        // Please KEEP this table SORTED ALPHABETICALLY! We do a binary search on this array.
+        private static InternalEncodingDataItem[] encodingDataTable = new[]
```

Comment:
OK - I wasn't sure on the implication of using more memory or not.  I'll make the switch.